### DISCUSSION TITLE
auto rebuild of dependent packages?

### DISCUSSION THREAD

**Discussion by rizsotto** [Homebrew/discussions#44] [Everyday usage] [answered]
Hey Folks,

I'm just using Homebrew in CI environment. To make sure my code works on Mac, because I don't own a Mac myself. Thanks for the great tool, it really makes cross platform testing easier.

Although I run into a problem a few times: I use gRPC which depends on other packages. And if the dependency updated the dependent packages are not rebuilt. Which cause my build to fail. (Which is confusing in a CI environment.)

My question, is there any automation implemented to re-build the dependent packages after a version bump?

**Comment by SMillerDev** [comment, marked as answer]
No, since there is no way of knowing if it needs to be rebuild. For homebrew-core `brew-test-bot` will tell us if there's linkage issues, which is usually a good indicator of a required rebuild. After that we ask the contributor to add `revision 1` to those formula.

**Reply by rizsotto** [reply]
Thanks for your answer @SMillerDev . I was thinking different approaches to deal with this kind of errors:

- Force to rebuild all dependent packages and bump up the revision. It gets rid of this kind of errors for sure. But it may do not necessary rebuilds, for packages which depends on many other packages. (I know that Nix is working like this.)
- Have a canary channel for packages. Where you can have time to capture this bugs before you hit production. (Fedora, Arch does this)
- Improve test coverage. Maybe the extend the signals for the test bot to trigger package rebuild.

Is this a frequent problem for Homebrew users? I'm using this system since June, but were bitten by this problem twice already.

**Reply by jonchang** [reply]
> Force to rebuild all dependent packages and bump up the revision. It gets rid of this kind of errors for sure. But it may do not necessary rebuilds, for packages which depends on many other packages. (I know that Nix is working like this.)

We already do this when our CI detects errors for packages in Homebrew/core.

> Have a canary channel for packages. Where you can have time to capture this bugs before you hit production. (Fedora, Arch does this)

This is unlikely to happen since we are already pretty much at our limits in terms of CI / maintainer capacity. 

> Improve test coverage. Maybe the extend the signals for the test bot to trigger package rebuild.

We always welcome improvements to tests, especially if they help us catch bugs when dependencies upgrade versions.

> Is this a frequent problem for Homebrew users? I'm using this system since June, but were bitten by this problem twice already.

If this is happening with packages in Homebrew/core, then it is a Homebrew problem and you should file issue reports in the appropriate repository. If this is for a package you're building in your own repository then the only solution currently is to watch for changes in Homebrew and rebuilding when appropriate, perhaps via a scheduled CI workflow. Homebrew is built on top of Git, so you can use anything that interfaces with Git to watch for the necessary changes.

**Reply by rizsotto** [reply]
Thanks @jonchang . Yes, this is a Homebrew/core problem, and I'll report it in a ticket.

But I raised this question here, because it seems to me this might be a frequent issue. (I might be wrong, but 2 cases in 4 months with the same package seems frequent to me.)
